File: ElementaryCellularAutomaton.java
// Created by: Thomas Conner

/* 
Elementary Cellular Automaton created without using arrays

This program is incomplete as I will stil need to implement "wrapping" using arithmetic operations

FOR STUDENTS:
If you are a student, do NOT copy and paste this code on a programming assignment. Your
teacher/professor will not expect you to use bitwise functions like this.

YOU WILL GET CAUGHT CHEATING

*/

package com.mycompany.main;

import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        
        //Probably not the best way to perform rudimentary input validation, but input validation
        //was sort of an after-thought with this program
        
        
        int rule = -1;
        while(rule < 0 || rule > 255){
            System.out.print("Enter rule (0-255): ");
            rule = scan.nextInt();
        }
        
        int numGens = -1;
        while(numGens<=0){
            System.out.print("Enter num generations: ");
            numGens = scan.nextInt();
        }
        
        int cur = 1<<16; //This makes it easier to change starting pos intead of putting an int literal
        
        for(int i = 0; i <= numGens; i++){
            drawRow(cur);
            cur = calcNextRow(cur, rule);
        }
        scan.close();
    }
    
    //Calculates the binary string representing the next row in the cellular automaton
    
    public static int calcNextRow(int prev, int rule){
        int cur = 0;
        
        for(int i = 0; i < 31; i++){
            cur=cur<<1;
            if((rule&(1<<((prev&(7<<29))>>>29)))!=0)
                cur++;
            int temp = prev&(3<<30)>>>30;

            prev=prev<<1 + temp;
        }
        return cur;
    }
    
    //Draws the given bit string in ascii characters
    
    public static void drawRow(int row){
        for(int i = 31; i >=0; i--){
            if((row&(1<<i)) == 1<<i){
                System.out.print("$");
            }
            else{
                System.out.print(".");
            }
        }
        System.out.println();
    }
}

File: ElementaryCellularAutomaton.cpp
#include <iostream>

unsigned int calcNextRow(unsigned int, unsigned int);
void drawRow(unsigned int);

int main() {
    unsigned int rule = 256;
    unsigned int numGens = 0;


    while(rule>255) {
        std::cout << "Enter rule (0-255): " << std::endl;
        std::cin >> rule;
    }

    while(numGens <= 0){
        std::cout << "Enter num generations: ";
        std::cin >> numGens;
    }

    unsigned int cur = 1<<16;

    for(int i = 0; i <= numGens; i++){
        drawRow(cur);
        cur = calcNextRow(cur, rule);
    }

    return 0;
}

unsigned int calcNextRow(unsigned int prev, unsigned int rule){
    unsigned int cur = 0;

    if((rule&(1<<(((prev&(3<<30))>>30) + ((prev&1)<<2)))) !=0){
        cur++;
    }

    for(int i = 0; i < 31; i++){
        cur=cur<<1;
        if((rule&(1<<((prev&(7<<29))>>29)))!=0)
            cur++;
        prev=(prev<<1)+((prev&(1<<31))>>31);
    }

    return cur;
}

void drawRow(unsigned int row){
    for(int i = 31; i >=0; i--){
        if((row&(1<<i)) == 1<<i){
            std::cout << "$";
        }
        else{
            std::cout << ".";
        }
    }
    std::cout << "\n";
}
